test/gateway/value-types.js

```diff
@@ -0,0 +1,410 @@
+    }
+    
+    type PostEdge {
+      node: Post!
+    }
+
+    type Post @key(fields: "id") {
+      id: ID!
+    }
+`
+
+const postsData = {
+  data: {
+    posts: {
+      pageInfo: {
+        hasNextPage: false
+      },
+      edges: [
+        {
+          node: {
+            id: '1'
+          }
+        }
+      ]
+    }
+  }
+}
+
+const postResolvers = {
+  Post: {
+    __resolveReference: (post, args, context, info) => {
+      return { id: '1' }
+    }
+  },
+  Query: {
+    posts: (root, args, context, info) => {
+      return postsData.data.posts
+    }
+  }
+}
+
+const postsQuery = `
+  query {
+    posts {
+      pageInfo {
+        hasNextPage
+      }
+      edges {
+        node {
+          id
+        }
+      }
+    }
+}`
+
+const commentSchema = `
+    extend type Query {
+      comments: CommentConnection!
+    }
+    
+    type CommentConnection {
+      pageInfo: PageInfo!
+      edges: [CommentEdge!]!
+    }
+    
+    type PageInfo {
+      hasNextPage: Boolean
+    }
+    
+    type CommentEdge {
+      node: Comment!
+    }
+
+    type Comment @key(fields: "id") {
+      id: ID!
+    }
+`
+
+const commentsData = {
+  data: {
+    comments: {
+      pageInfo: {
+        hasNextPage: false
+      },
+      edges: [
+        {
+          node: {
+            id: '1'
+          }
+        }
+      ]
+    }
+  }
+}
+
+const commentResolvers = {
+  Comment: {
+    __resolveReference: (post, args, context, info) => {
+      return { id: '1' }
+    }
+  },
+  Query: {
+    comments: (root, args, context, info) => {
+      return commentsData.data.comments
+    }
+  }
+}
+
+const commentsQuery = `
+  query {
+    comments {
+      pageInfo {
+        hasNextPage
+      }
+      edges {
+        node {
+          id
+        }
+      }
+    }
+}`
+
+test('Should be able to query with value types', async (t) => {
+  const [userService, userServicePort] = await createService(t, userSchema, userResolvers)
+  const [postService, postServicePort] = await createService(t, postSchema, postResolvers)
+  const [commentService, commentServicePort] = await createService(t, commentSchema, commentResolvers)
+
+  const gateway = Fastify()
+
+  t.teardown(async () => {
+    await gateway.close()
+    await postService.close()
+    await userService.close()
+    await commentService.close()
+  })
+
+  gateway.register(GQL, {
+    gateway: {
+      services: [{
+        name: 'user',
+        url: `http://localhost:${userServicePort}/graphql`
+      }, {
+        name: 'post',
+        url: `http://localhost:${postServicePort}/graphql`
+      }, {
+        name: 'comment',
+        url: `http://localhost:${commentServicePort}/graphql`
+      }]
+    }
+  })
+
+  const usersRes = await gateway.inject({
+    method: 'POST',
+    headers: {
+      'content-type': 'application/json'
+    },
+    url: '/graphql',
+    body: JSON.stringify({
+      query: usersQuery
+    })
+  })
+
+  t.same(JSON.parse(usersRes.body), usersData)
```

Comment:
```suggestion
  t.same(usersRes.json(), usersData)
```

and everywhere else.